allennlp/modules/self_attentive_sentence_encoder.py

```diff
@@ -0,0 +1,123 @@
+
+        super(SelfAttentiveSentenceEncoder, self).__init__()
+        self._attention_size = attention_size
+        self._num_attention_heads = num_attention_heads
+        self._input_dim = input_dim
+        self._forbenius_regularization = forbenius_regularization
+        self._regularization_coeffecient = regularization_coeffecient
+
+        self._linear_inner = torch.nn.Linear(input_dim, attention_size, bias=False)
+        self._linear_outer = torch.nn.Linear(attention_size, num_attention_heads, bias=False)
+
+    def init_weights(self) -> None:
+        initrange = 0.1
+        self.linear_first.weight.data.uniform_(-initrange, initrange)
+        self.linear_second.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, # pylint: disable=arguments-differ
+                inputs: torch.Tensor,
+                mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        """
+        Parameters
+        ----------
+        inputs : ``torch.FloatTensor``, required.
+            A tensor of shape (batch_size, timesteps, input_dim)
+        mask : ``torch.FloatTensor``, optional (default = None).
+            A tensor of shape (batch_size, timesteps).
+
+        Returns
+        -------
+        A dictionary of outputs containing the following items:
+        representation : ``torch.FloatTensor``
+        The final representation produced by attention of the shape (batch_size, input_dim).
+        penalty : ``torch.FloatTensor``
+        The forbenius norm based regularization penalty.
+        attention : ``torch.FloatTensor``
+        The values of attention corrosponding to the different attention heads having
+         the shape (batch_size, num_attention_heads, time_steps).
+        """
+        # Shape (batch_size, timesteps, attention_size)
+        attention_matrix = torch.tanh(self._linear_inner(inputs))
+        # Shape (batch_size, timesteps, num_attention_heads)
+        attention_vector = self._linear_outer(attention_matrix)
+        # Shape (batch_size, timesteps, num_attention_heads)
+        if isinstance(mask, torch.Tensor):
+            mask = mask.unsqueeze(2)  # For unsqueezing mask to have three dimensions
+
+        attention = masked_softmax(attention_vector, mask, dim=1)
+
+        batch_size = inputs.shape[0]
+        outputs = {"attention": attention}
+
+        if self._forbenius_regularization:
+            # Shape (batch_size, num_attention_heads, timesteps)
+            attention_transpose = attention.transpose(1, 2)
+
+            # Done to ensure that constant identity matrix is also created on the device model is running
+            if isinstance(attention, torch.FloatTensor):
+                identity = torch.eye(attention.size(1))
+            else:
+                identity = torch.eye(attention.size(1)).cuda()
+
+            # Shape (batch_size, timesteps, timesteps)
+            identity = identity.unsqueeze(0).expand(batch_size, attention.size(1), attention.size(1))
```

Comment:
you use `attention.size(1)` in a lot of places, might be clearer if you assigned it to a variable

```
num_timesteps = attention.size(1)
```

and then used that